### PR TITLE
CR 21180-C++ Camera app -- core changes

### PR DESCRIPTION
**Pull request by humbletim**
This PR contains the C++-side changes for the camera movement tablet app.

### New / Updated APIs:
* **`MyAvatar.lookAtSnappingEnabled`** (bool; default true)
  - New property for toggling automatic eye contact snapping on/off per Interface client.
  - Given any two avatars, both will now need this enabled (which it is by default) for snapping to apply.
  - The value is persisted to settings and also available through `Developer > Avatar > Enable LookAt Snapping`.
* **`Reticle.scale`** (float; default 1.0)
  - New property for configuring mouse cursor scaling.
  - HMD mode had this feature for a while and several of its undocumented(?) aspects now apply to Desktop mode too:
    - <kbd>Control-'</kbd> -- cycles through available cursors (including the operating system cursor -- which can be useful for screen captures since desktop recording software generally knows how to cope with the standard cursor).
    - <kbd>Control-Numpad-plus</kbd> / <kbd>Control-Numpad-minus</kbd>  )</kbd> -- dynamically scales the mouse cursor.
* **`MyAvatar.pitchSpeed`** / **`MyAvatar.yawSpeed`** (float; degrees/second)
  - New properties for configuring angular speeds from scripting.
  - These values were only available from a UI dialog before ("Settings > Avatar > Avatar Camera") and not actually persisted anywhere between restarts; now they are saved and restored to settings per usual.

---------

### LookAt Snapping:
  - Launch this PR's sandbox and two (A/B) instances of Interface.
  - Verify both fresh starts begin with `Developer > Avatar > Enable LookAt Snapping` enabled.
  - Enable `Developer > Avatar > Show Other Eye Vectors` and `Show My Eye Vectors` (on both instances).
  - Position yourselves to look at yourselves...
  - Verify that eye snapping works per usual and mutually.
  - Disable A's snapping by unchecking "Enable LookAt Snapping".
  - Verify neither A nor B now makes snapped eye contact.
  - Verify B's Eye Vectors are rendered in normal red/blue, while A's are now rendered in dark red/blue.  *(this should be the case from either instance -- ie: A's should be dark, whether local or remote)*.
  - Reverse the configuration so A is re-enabled and B is now disabled.
  - Verify that neither avatar snaps, A's Eye Vectors are normal red/blue, and B's are now dark red/blue.

### Pitch/Yaw Speed settings -- Desktop mode
  - Bring up the `Console...` and test these `> commands` one at a time (followed by <kbd>Enter</kbd>).
```javascript
// console commands
> MyAvatar.yawSpeed
< 120
> MyAvatar.pitchSpeed
< 90
> MyAvatar.pitchSpeed = 15
< 15
> MyAvatar.pitchSpeed
< 15
```
  - Open `Settings > Avatar > Avatar Camera`
    - Verify the "Camera pitch speed" field is `15`.
    - Set the "Camera yaw speed" field to `240`, and close the dialog with "Save Changes".

```javascript
// console commands
> MyAvatar.yawSpeed
< 240
```
  - Restart the application.
    - Verify the displayed pitch and yaw values have remained the same as before.

### Mouse cursor scaling -- Desktop mode
  - Launch this PR's interface.
  - Verify that <kbd>Control-Numpad-plus</kbd> makes the mouse cursor bigger and <kbd>Control-Numpad-minus</kbd> makes it smaller.
  - Open the `Console...` and enter these commands:
    - `Reticle.scale` [enter]
      - output should be a floating point number proportionate to the current mouse cursor size
    - `Reticle.scale = 2.0` [enter]
      - the visible mouse cursor should become twice normal size
    - `Reticle.scale = 0.25` [enter]
      - the visible mouse cursor should become quarter normal size

### Mouse cursor scaling -- HMD Mode
  - Launch this PR's interface and perform a quick smoke around features where the mouse cursor and UI lasers come into play.
  - Verify in general that HMD mouse cursor behavior is consistent with previous releases.

//

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/437/HighFidelity-Beta-PR10625-7cab0363a929fb86e7f13b9f552488d1acb856c4.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/437/HighFidelity-Beta-PR10625-7cab0363a929fb86e7f13b9f552488d1acb856c4.dmg

**Comment by OmegaHeron** [comment]
QA In progress.

**Comment by OmegaHeron** [comment]
Returning this to QA Pool - Am no longer able to compile Linux build that will work with Windows PR Installer -- proto mismatch invoked, though Linux build shows PR10625 for all its versioning.  Can't get a 2 person setup without it and now out of time.

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

[Using 2 desktops + 1 HMD-Vive]

"LookAt Snapping"
All **Passes**. No issues.

"Pitch/Yaw Speed settings -- Desktop mode"
All **Passes**, Console brings up the correct values before and after restarting/changes

"Mouse cursor scaling -- Desktop mode"
All **Passes*.


"Mouse cursor scaling -- HMD Mode"
Not sure if **Pass** or **Fail**?, I noticed now I can no longer see the mouse at all when in HMD, but I don't believe this is normal?
(I tested out cursor scale set from various numbers such as 0.25, 1, 2. But unfortunately it was not visible)

Please let me know if there is anything else I can do/provide to help :)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/888/HighFidelity-Beta-PR10625-fbeaa01f7db70dec523c76c649603d8c4c7f5529.exe

**Comment by marko8904** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/899/HighFidelity-Beta-PR10625-fbeaa01f7db70dec523c76c649603d8c4c7f5529.dmg

**Comment by thoys** [comment]
test this please

**Comment by marko8904** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/937/HighFidelity-Beta-PR10625-fbeaa01f7db70dec523c76c649603d8c4c7f5529.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/937/HighFidelity-Beta-PR10625-fbeaa01f7db70dec523c76c649603d8c4c7f5529.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1071/HighFidelity-Beta-PR10625-f6262ad54d41e3e05403a0f5f2002dc6bf210ae9.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1071/HighFidelity-Beta-PR10625-f6262ad54d41e3e05403a0f5f2002dc6bf210ae9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1283/HighFidelity-Beta-PR10625-9053531bda7d157b46e95c7070d5c65bdf907537.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1283/HighFidelity-Beta-PR10625-9053531bda7d157b46e95c7070d5c65bdf907537.exe

**Comment by themelissabrown** [comment]
@humbletim does this need tested again?

**Comment by Nex-Pro** [comment]
Testing this

**Comment by themelissabrown** [comment]
test this please

**Comment by Nex-Pro** [comment]
Will jump on this before or after the meeting depending on how long the build takes to arrive.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1457/HighFidelity-Beta-PR10625-9053531bda7d157b46e95c7070d5c65bdf907537.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1457/HighFidelity-Beta-PR10625-9053531bda7d157b46e95c7070d5c65bdf907537.exe

**Comment by Nex-Pro** [comment]
In progress

**Comment by Nex-Pro** [comment]
### Pass
apologies for delay in completion my Vive decided I only had one hand controller.
Refer to test plan spreadsheet. https://docs.google.com/spreadsheets/d/1VzTP5Wtn1eTsRz5kjvKq8Vuf2Ovxzu1eys8MbhUHVKs/edit#gid=912121214

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -1246,14 +1260,19 @@ void MyAvatar::updateLookAtTargetAvatar() {
                  ViewFrustum viewFrustum;
                  qApp->copyViewFrustum(viewFrustum);
 
+                 glm::vec3 viewPosition = viewFrustum.getPosition();
+                 glm::quat viewOrientation = viewFrustum.getOrientation();
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: `viewOrientation` is only ever used in the `if (qApp->isHMDMode())` block below.  Maybe move the declaration and initialization of `viewOrientation` into that scope.

File: libraries/avatars/src/AvatarData.cpp [hunk 2]
```diff
@@ -1510,7 +1512,9 @@ void AvatarData::processAvatarIdentity(const QByteArray& identityData, bool& ide
             >> identity.displayName
             >> identity.sessionDisplayName
             >> identity.isReplicated
-            >> identity.avatarEntityData;
+            >> identity.avatarEntityData
+            >> identity.lookAtSnappingEnabled
```
Comment (AndrewMeadows) [inline libraries/avatars/src/AvatarData.cpp]: This PR is packing new data into the Avatar's identity packet.  I think that means this PR is introducing a change to the network protocol.  The packet version must be updated in `versionForPacketType()`, under the `AvatarIdentity` case.  PR-10956 provides a recent example a project that required a protocol change.

File: libraries/avatars/src/AvatarData.h [hunk 3]
```diff
@@ -324,6 +324,7 @@ class AvatarData : public QObject, public SpatiallyNestable {
 
     Q_PROPERTY(glm::vec3 position READ getPosition WRITE setPosition)
     Q_PROPERTY(float scale READ getTargetScale WRITE setTargetScale)
+    Q_PROPERTY(float density READ getDensity)
```
Comment (AndrewMeadows) [inline libraries/avatars/src/AvatarData.h]: Avatar density is not yet writable: it is hard coded in C++ to the density of water.  Eventually we'll need to expose it to script for reading (and writing).  I'm wondering: Why we would want to expose it for reading now?

Note, we don't yet expose mass or volume on the `MyAvatar` object either, however the mass of the avatar will increase or shrink as its scale is made larger and smaller respectively -- the change is sometimes evident in how the avatar collides with dynamic objects.

File: interface/src/Application.cpp [hunk 4]
```diff
@@ -4153,10 +4156,13 @@ void Application::updateMyAvatarLookAtPosition() {
         }
     } else {
         AvatarSharedPointer lookingAt = myAvatar->getLookAtTargetAvatar().lock();
-        if (lookingAt && myAvatar.get() != lookingAt.get()) {
+        bool haveLookAtCandidate = lookingAt && myAvatar.get() != lookingAt.get();
+        auto avatar = static_pointer_cast<Avatar>(lookingAt);
+        bool mutualLookAtSnappingEnabled = avatar && avatar->getLookAtSnappingEnabled() && myAvatar->getLookAtSnappingEnabled();
+        if (haveLookAtCandidate && mutualLookAtSnappingEnabled) {
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: So we only fall into the code below if `MyAvatar` has `LookAtSnappingEnabled`  **AND**  the avatar I might look at also has `LookAtSnappingEnabled`.  Why would MyAvatar's behavior require the other avatar to also have it enabled?

Comment (AndrewMeadows) [inline interface/src/Application.cpp]: BTW, how does the `lookAt` stuff work?  Does `MyAvatar` compute what it is looking at and relay that info through the `AvatarMixer`? or does the local interface compute `lookAt` for other avatars?  If it is the latter then yes, we need to know whether the other avatars have the `lookAt` feature enabled, but if it is the former then maybe we don't actually need the setting's enable/disable state relayed through the `AvatarMixer`... and don't need to change the `AvatarIdentity` packet protocol.

Comment (AndrewMeadows) [inline interface/src/Application.cpp]: Ok I see that this behavior is explicitly listed as intended in the PR description.  What this means is:

If I disable LookAtSnapping on my client then not only will MyAvatar not snap its eyes but as far as my scene is concerned so will all other avatars.  Meanwhile, someone else with LookAtSnapping enabled will see everyone else snapping except me.

Comment (humbletim) [inline interface/src/Application.cpp]: Yup, that sounds about right.  Just to be clear -- this setting only affects situations involving `MyAvatar` directly.  Even when I disable `LookAtSnapping`, *other avatar pairs* in the scene will still be observed snapping onto each other when making eye contact.

File: libraries/avatars/src/AvatarData.h (hunk 3 above)
Comment (humbletim) [inline libraries/avatars/src/AvatarData.h]: This property emerged as part of troubleshooting thrust calculations in context of the new density logic.  I don't see any need for it from a scripting either yet, but thought maybe leaving it in place could serve as a nice breadcrumb / starting point for whoever comes back later to make things read/write?  No problem removing though if that'd make more sense for now.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1761/HighFidelity-Beta-PR10625-d9de0a8687d9ef70aa42fb9605395fc2e8ee3e13.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1761/HighFidelity-Beta-PR10625-d9de0a8687d9ef70aa42fb9605395fc2e8ee3e13.exe

File: libraries/avatars/src/AvatarData.cpp (hunk 2 above)
Comment (humbletim) [inline libraries/avatars/src/AvatarData.cpp]: ... packet version updated!

shortcut to this PR's PacketHeader.* changes: https://github.com/highfidelity/hifi/pull/10625/files#diff-9d0aa4edd54d124eef1e33378b6affed

**Comment by humbletim** [comment]
@AndrewMeadows I think all of the requested changes have been done -- could you please confirm?

**Comment by AndrewMeadows** [comment]
Ok this is basically ready to go, but let's coordinate with @cozza13  and/or other devs and merge this in the daytime when we're ready to accept a protocol version change on the dev domains.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2129/HighFidelity-Beta-PR10625-3bcbec87e0f7e0a2f312708fe5ac7f2fb782f21e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2129/HighFidelity-Beta-PR10625-3bcbec87e0f7e0a2f312708fe5ac7f2fb782f21e.exe

**Comment by humbletim** [comment]
bump

**Comment by 1st-BrainStormer** [comment]
I hope this goes in soon because I can never find my mouse curson on a 4k Monitor because hifi ignores the system cursor size and shrinks the cursor.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2739/HighFidelity-Beta-PR10625-9348048387c9a83ee1fdb1e6c30faaacc13765fe.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2739/HighFidelity-Beta-PR10625-9348048387c9a83ee1fdb1e6c30faaacc13765fe.exe

**Comment by thoys** [comment]
test this please

**Comment by cain-kilgore** [comment]
Testing.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2745/HighFidelity-Beta-PR10625-9348048387c9a83ee1fdb1e6c30faaacc13765fe.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2745/HighFidelity-Beta-PR10625-9348048387c9a83ee1fdb1e6c30faaacc13765fe.dmg

**Comment by cain-kilgore** [comment]
# Test Failed

## Lookat Snapping
Test | Result
--- | ---
Launch this PR's sandbox and two (A/B) instances of Interface. | Pass
Verify both fresh starts begin with Developer > Avatar > Enable LookAt Snapping enabled. | Pass
Enable Developer > Avatar > Show Other Eye Vectors and Show My Eye Vectors (on both instances). | Pass
Position yourselves to look at yourselves... | Pass
Verify that eye snapping works per usual and mutually. | Pass
Disable A's snapping by unchecking "Enable LookAt Snapping". | Pass
Verify neither A nor B now makes snapped eye contact. | Pass
Verify B's Eye Vectors are rendered in normal red/blue, while A's are now rendered in dark red/blue. (this should be the case from either instance -- ie: A's should be dark, whether local or remote). | Pass
Reverse the configuration so A is re-enabled and B is now disabled. | Pass
Verify that neither avatar snaps, A's Eye Vectors are normal red/blue, and B's are now dark red/blue. | Pass

## Pitch / Yaw Speed Settings
Test | Result
--- | ---
Bring up the Console... and test these > commands one at a time (followed by Enter). |
MyAvatar.yawSpeed | Pass
120 | Pass (100)
MyAvatar.pitchSpeed | Pass
90 | Pass
MyAvatar.pitchSpeed = 15 | Pass
15 | Pass
MyAvatar.pitchSpeed | Pass
15 | Pass
Open Settings > Avatar > Avatar Camera | Pass
Verify the "Camera pitch speed" field is 15. | Pass
Set the "Camera yaw speed" field to 240, and close the dialog with "Save Changes". | Pass
MyAvatar.yawSpeed | Pass
240 | Pass
Restart the application. | Pass
Verify the displayed pitch and yaw values have remained the same as before. | Pass

## Mouse Cursor Scaling (Desktop)
Test | Result
--- | ---
Launch this PR's interface. | Pass
Verify that Control-Numpad-plus makes the mouse cursor bigger and Control-Numpad-minus makes it smaller. | Pass
Open the Console... and enter these commands: | 
Reticle.scale [enter] | Pass
output should be a floating point number proportionate to the current mouse cursor size | Pass
Reticle.scale = 2.0 [enter] | Pass
the visible mouse cursor should become twice normal size | Pass
Reticle.scale = 0.25 [enter] | Pass
the visible mouse cursor should become quarter normal size | Pass

## Mouse Cursor Scaling (HMD/Vive)
Test | Result
--- | ---
Launch this PR's interface and perform a quick smoke around features where the mouse cursor and UI lasers come into play. | Pass
Verify in general that HMD mouse cursor behavior is consistent with previous releases. | Fail (See notes)

## Notes
I wasn't able to see the mouse cursor in HMD. ***If this is expected*** then this is a solid pass from me.

###### Intel i5 7600K @ 4Ghz, 16GB DDR4 2133Mhz Corsair Vengeance, Asus Strix GTX 1050Ti 3GB, HTC Vive

**Comment by humbletim** [comment]
@cain-kilgore re: HMD cursor fail -- So the current beta works differently (ie: doesn't lose the mouse cursor in the same situation)?

I heard another user mention the HMD mouse cursor disappearing earlier this week -- but in that case it would have been on the the current or just prior beta.

**Comment by marko8904** [comment]
build this please

**Comment by cain-kilgore** [comment]
Hey @humbletim: You're right - I wasn't able to see the cursor in the master either. Originally you could and I don't know why this was removed. Your PR is a Pass from me :)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2758/HighFidelity-Beta-PR10625-9348048387c9a83ee1fdb1e6c30faaacc13765fe.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2758/HighFidelity-Beta-PR10625-9348048387c9a83ee1fdb1e6c30faaacc13765fe.dmg

**Pull request merged**